ansible/roles/test/files/ptftests/inner_hash_test.py

```diff
@@ -0,0 +1,505 @@
+from ptf.base_tests import BaseTest
+from ptf.mask import Mask
+from ptf.testutils import *
+
+import fib
+import lpm
+
+class InnerHashTest(BaseTest):
+
+    #---------------------------------------------------------------------
+    # Class variables
+    #---------------------------------------------------------------------
+    DEFAULT_BALANCING_RANGE = 0.25
+    BALANCING_TEST_TIMES = 625
+
+    _required_params = [
+        'fib_info',
+        'src_ports',
+        'inner_src_ip_range',
+        'inner_dst_ip_range',
+        'outer_src_ip_range',
+        'outer_dst_ip_range'
+    ]
+
+
+    def __init__(self):
+        '''
+        @summary: constructor
+        '''
+        BaseTest.__init__(self)
+        self.test_params = test_params_get()
+        self.check_required_params()
+
+
+    def check_required_params(self):
+        for param in self._required_params:
+            if param not in self.test_params:
+                raise Exception("Missing required parameter {}".format(param))
+
+
+    def setUp(self):
+        '''
+        @summary: Setup for the test
+        '''
+        self.dataplane = ptf.dataplane_instance
+
+        self.fib = fib.Fib(self.test_params['fib_info'])
+        self.router_mac = self.test_params['router_mac']
+
+        inner_src_ip_range = [unicode(x) for x in self.test_params['inner_src_ip_range'].split(',')]
+        inner_dst_ip_range = [unicode(x) for x in self.test_params['inner_dst_ip_range'].split(',')]
+        self.inner_src_ip_interval = lpm.LpmDict.IpInterval(ip_address(inner_src_ip_range[0]), ip_address(inner_src_ip_range[1]))
+        self.inner_dst_ip_interval = lpm.LpmDict.IpInterval(ip_address(inner_dst_ip_range[0]), ip_address(inner_dst_ip_range[1]))
+
+        outer_src_ip_range = [unicode(x) for x in self.test_params['outer_src_ip_range'].split(',')]
+        outer_dst_ip_range = [unicode(x) for x in self.test_params['outer_dst_ip_range'].split(',')]
+        self.outer_src_ip_interval = lpm.LpmDict.IpInterval(ip_address(outer_src_ip_range[0]), ip_address(outer_src_ip_range[1]))
+        self.outer_dst_ip_interval = lpm.LpmDict.IpInterval(ip_address(outer_dst_ip_range[0]), ip_address(outer_dst_ip_range[1]))
+
+        self.hash_keys = self.test_params.get('hash_keys', ['src-ip', 'dst-ip', 'src-port', 'dst-port'])
+        self.src_ports = self.test_params['src_ports']
+        self.symmetric_hashing = self.test_params.get('symmetric_hashing', False)
+        self.outer_dst_ip = self.outer_dst_ip_interval.get_first_ip()
+
+        self.next_hop = self.fib[self.outer_dst_ip]
+        self.exp_port_list = self.next_hop.get_next_hop_list()
+        assert (len(self.exp_port_list) > 1)
+        for exp_port in self.exp_port_list:
+            assert exp_port not in self.src_ports
+
+        self.balancing_range = self.test_params.get('balancing_range', self.DEFAULT_BALANCING_RANGE)
+        self.balancing_test_times = self.test_params.get('balancing_test_times', self.BALANCING_TEST_TIMES)
```

Comment:
@anish-n please add logging for input params

Reply:
If the test fails, the error will return the exact set of test params passed in as part of standard PTF test failure behavior, it seems to me that suffices, if not, can you add some info as to how it would help?